Add full theming documentation in English (#24)
applidget/event-theme-maker: lib/new_repository/template/docs/theming.md

# Eventmaker website theme squeleton

This is a starting point for an Eventmaker Website theme.

## Folder structure

    ├── themes
    │   ├── standard
    |   |   ├── assets
    |   |   |   ├── // asset file
    |   |   |   ├── ...
    |   |   ├── config
    |   |   |   ├── main_settings_schema.json
    |   |   |   ├── presets.en.json
    |   |   |   ├── presets.fr.json
    |   |   |   ├── sections-groups.json
    |   |   |   ├── translations.json
    |   |   ├── layouts
    |   |   |   ├── // layout file
    |   |   |   ├── ...
    |   |   ├── sections
    |   |   |   ├── // section file
    |   |   |   ├── ...
    |   |   ├── snippets
    |   |   |   ├── // snippet file
    |   |   |   ├── ...
    |   |   ├── templates
    |   |   |   ├── // file for error pages
    |   |   |   ├── ...
    |   |   ├── package.json
    |   |   ├── specs.yml
    └── README.md

# Working on a theme

> The language used is the "Liquid" developed by Shopify, a doc is available [here](https://shopify.github.io/liquid).

Reviewer: ```suggestion
The language used is the "Liquid" developed by Shopify. Documentation is available [here](https://shopify.github.io/liquid).
```

You have access to the parameters of the URL with the liquid variable `url_params` (ex: with https://mywebsite.eventmaker.io/index?superParams=true to know the value of `superParams` we will do `url_params.superParams` which will return a String `true`).

# Config

> ## spec.yml

Reviewer: ```suggestion
## specs.yml
```

## package.json

This file allows you to configure the execution of the Yarn script as well as to install theme specific dependencies during Webpack builds.

## config\main_settings_schema.json

This file allows you to create the general parameters of the site (social networks, title color, background color, text color, section spacing, font, etc...). Everything that is not specific to a single section and/or that must appear in the layout must exist here.

To reuse these settings in the HTML you need to do `settings.id_of_setting` (e.g. to call the text color of the website: `settigns.color_body_text`).

## config\presets.fr.json and presets.en.json

These are the default settings when creating a website with this theme.

In order to have directly a functional website you can choose the default settings, such as site colors, existing pages and sections and their content of each.

## config-sections-groups.json

This file allows you to create different groups of sections in order to order them. When adding a section via the customization interface, the different sections are then grouped together (basic, media, registration, etc.).

> ## configtranslations.json

Reviewer: typo I guess ?

This file allows you to configure all the texts which are not customizable, but which must be translated into the language of the website.

Each language is represented by its ISO key (en for english, fr for french, it for italian, etc...) and each translation is in the form `"key": "value"`.

> Translations can be used anywhere on the site by doing `t.key_of_translation`.

Reviewer: ```suggestion
Translations can be used in liquid files by doing `t.key_of_translation`.
```

> If you want to have dynamic content in a translation it will look like this: `"key": "My name is %{ variable_text }"`. For this to work there must also be a liquid variable in the file where this translation is called whose value is the changing part (`{% capture variable_text %}{{ guest.first_name }}{% endcapture %}`).

Reviewer: ```suggestion
If you need dynamic content in a translation use the `%{}` placeholder: `"key": "My name is %{ variable_text }"`. The variable text must be a liquid variable. For example:

```
{% capture variable_text %}{{ guest.first_name }}{% endcapture %}`).
```
```

# Assets

> This folder contains the resources of the theme Css and JS files they are compiled by Webpacker and can make dynamic calls to files contained in other folder of the repo by @import (Css) or import (JS).

Reviewer: ```suggestion
This folder contains the resources of the theme CSS and JS files. They are compiled by Webpack and can make dynamic calls to files contained in other folder of the repo by @import (CSS) or import (JS).
Webpack entry point is `main.js` and `main.scss`.
```

# Layout

> Layouts are skeletons for each page of the site. It contains the elements that must be present at each time (CSS, Js, Favicon, etc. ...).

Reviewer: ```suggestion
Layouts are skeletons for each page of the site. It contains the elements that must be present on each page (CSS, JS, Favicon, etc. ...).
```

# Sections

> The sections are the elements that can be added to a page via the customization interface of the website.

Reviewer: ```suggestion
The sections are the elements that can be added to a page via the builder.
```

> Each section consists of two distinct parts: the HTML code and the schema.

Reviewer: ```suggestion
Each section consists of two distinct parts: the HTML/liquid code and the schema.
```

## HTML code of a section

In the HTML code you can use the general parameters `settings.id_of_the_setting`. But also and especially the settings of the section created in the schema: `section.settings.id_of_the_section_setting`.

## Schema of a section

The schema gives us information about the section and allows us to create the options we want to customize the section.

    {% schema %}
      {
        ...
      }
    {% endschema %}

> The first elements of the diagram are general information about the section.

Reviewer: ```suggestion
The first elements are generally information about the section.
```

    {% schema %}
      {
        "name": "English title - show on the website builder",
        "name_translations": { "fr": "Titre en français - afficher dans l'interface de personnalisation du site web" },
        "icon": "fa fa-font", // Font-awesome classes for the icon display in the customization interface. https://fontawesome.com/v4/icons/
        "hidden_from_user": false,
        "group_key": "basic", // In which group the section will be displayed
        "group_rank": 1000, // Its order in the group
        "settings": [
          ...
        ]
      }
    {% endschema %}

## Settings

> 
> What are the `settings` looks like?

Reviewer: ```suggestion
```

- `info` information display under the parameter, if the back office is in English.
- `info_translations` works the same way as `label_translation`.
- `default` is the default value filled in if the back office is in English.
- `default_translations` works the same way as `label_translation`.

#### Specificities

Some setting types need more than a `label`, an `id` and a `type` to work.

- `select` : settings of type _select_ need options to work. The empty option is generated automatically.

Example:

    {
      "type": "select",
      "id": "text_size",
      "label": "Text size",
      "label_translations": { "fr": "Taille du texte" },
      "default": "medium",
      "options": [
        {
          "label": "Medium",
          "label_translations": { "fr": "Moyen" },
          "value": "medium"
        },
        {
          "label": "Large",
          "label_translations": { "fr": "Gros" },
          "value": "large"
        }
      ]
    }

- `checkbox`: its default value will be a boolean (`true` / `false`), not a String.
- `color` : its default value must be a hexadecimal code.

#### Display condition - Show_if

> To each setting it is possible to add a display condition (`show_if`).

Reviewer: ```suggestion
For each setting it is possible to add conditional display (`show_if`).
```

      "label_translations": { "fr": "Appliquer un style personnalisé" },
      "default": false
    },
    {
      "type": "color",
      "id": "button_color",
      "label": "Button color",
      "label_translations": { "fr": "Couleur du bouton" },
      "default": "#000",
      "show_if": [
        [
          {
            "source_id": "enable_custom_css",
            "operator": "==",
            "value": true
          }
        ]
      ]
    }

It is possible to make multiple conditions with AND :

If `checkbox_1` == true AND `checkbox_2` == true

    "show_if": [
      [
        {
          "source_id": "checkbox_1",
          "operator": "==",
          "value": true
        },
        {
          "source_id": "checkbox_1",
          "operator": "==",
          "value": true
        }
      ]
    ]

But also ORs:

If `checkbox_1` == true OR `checkbox_2` == true

    "show_if": [
      [
        {
          "source_id": "checkbox_1",
          "operator": "==",
          "value": true
        }
      ],
      [
        {
          "source_id": "checkbox_1",
          "operator": "==",
          "value": true
        }
      ]
    ]

### Blocks

> Blocks are elements that the user of the customization interface can add within the available limit (if `max_blocks` is defined). This allows to create elements of the same nature several times (e.g. in the section `feature-colums.liquid` the blocks are used to create the different columns. We can make 1, 2, 4, or more and manage their display without making several sections).

Reviewer: ```suggestion
Blocks are elements that the user of the builder can add within the available limit (if `max_blocks` is defined). This allows to create elements of the same nature several times (e.g. in the section `feature-colums.liquid` the blocks are used to create the different columns. We can make 1, 2, 4, or more and manage their display without making several sections).
```

            "name": "Blocks",
            "name_translations": { "fr": "Blocs" },
            "add_block": "Text for a new block",
            "add_block_translations": { "fr": "Texte pour un Nnuveau bloc" },
            "remove_block": "Text to remove this block",
            "remove_block_translations": { "fr": "Texte pour retirer ce bloc" },
            "display_property": "id_of_block_settings",
            "settings": [
              ...
            ]
          }
        ],
        "presets": [
          {
            "name": "Name of the section",
            "name_translations": { "fr": "Nom de la section" },
            "blocks": [
              {
                "type": "text_block"
              },
              {
                "type": "text_block"
              },
              {
                "type": "text_block"
              }
            ]
          }
        ]
      }
    {% endschema %}

# Snippets

Snippets are code fragments that can be used in several sections or layouts. To avoid having several times the same code in different sections, you can include a snippet.

To do this you just have to do `{% include "filename_of_the_snippet" %}`.

> You can also pass parameters to this snippet: `{% include "filename_of_the_snippet", section_settings: section.settings %}`.
> And so in the snippet you have access to this `section_settings` parameter.

Reviewer: ```suggestion
You can also pass parameters to this snippet: `{% include "filename_of_the_snippet", section_settings: section.settings %}` and access them in the snippet with `section_settings`.
```

Unlike sections, snippets do not have a schema.

# Templates

The `templates` folder contains page templates for different cases (normal operation, 401 error, 404 error, Offline, etc...).

# Liquid variables

You have access to different Liquid variables:

## Global variables

> Here is the list of Liquid variables available on the website.

Reviewer: ```suggestion
Here is the list of Liquid variables available:
```
